### PR TITLE
[SQL-DS-CACHE-24][POAE7-893] HCFS library for cache locality in APE

### PR DESCRIPTION
**Pull request by iyupeng**
## What changes were proposed in this pull request?

add a HCFS (Hadoop Compatible File System) library which wraps HDFS data accessing and provides locations of cached data.

## How was this patch tested?

integration tests

### CODE REVIEW

File: oap-ape/ape-java/ape-hcfs/pom.xml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>ape-hcfs</artifactId>
+    <version>0.0.1-SNAPSHOT</version>
+    <packaging>jar</packaging>
+    <name>A hadoop compatible file system that wraps HDFS API and ape cache locality</name>
+
+    <properties>
+        <hadoop.version>2.4.1</hadoop.version>
```
Comment (winningsix) [inline oap-ape/ape-java/ape-hcfs/pom.xml:20]: Move ```Hadoop version``` to parent POM please.

**Comment by iyupeng** [comment]
@winningsix @jikunshang @xieqi Please take a review, thanks.

File: oap-ape/ape-java/ape-hcfs/src/main/java/com/intel/oap/fs/hadoop/ape/hcfs/CacheAwareFileSystem.java [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+    private org.apache.hadoop.fs.FileSystem hdfs;
+
+    private String locationPolicy;
+
+    @Override
+    public void initialize(URI name, Configuration conf) throws IOException {
+        uri = name;
+        hdfs = new DistributedFileSystem();
+        hdfs.initialize(name, conf);
+        LOG.info("fs initialized.");
```
Comment (winningsix) [inline oap-ape/ape-java/ape-hcfs/src/main/java/com/intel/oap/fs/hadoop/ape/hcfs/CacheAwareFileSystem.java:68]: Better call it "cache aware scheduler is initialized"

File: oap-ape/ape-java/ape-hcfs/src/main/java/com/intel/oap/fs/hadoop/ape/hcfs/CacheAwareFileSystem.java [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+    @Override
+    public BlockLocation[] getFileBlockLocations(Path path, long start, long len) throws IOException {
+        if (path == null) {
+            throw new NullPointerException();
+        }
+        LOG.debug("getFileBlockLocations with: {}, start: {}, len: {}", path.toString(), start, len);
+
+        List<BlockLocation> result = new ArrayList<>();
+
+        if (start >= 0 && len > 0) {
```
Comment (winningsix) [inline oap-ape/ape-java/ape-hcfs/src/main/java/com/intel/oap/fs/hadoop/ape/hcfs/CacheAwareFileSystem.java:94]: ```if (start >= 0 && len > 0)``` should we throw an exception if must matched case?
BTW, better to 
```
if (start < 0 || len < 0) {
  return result.toArray(new BlockLocation[0]);
}
```

**Review by winningsix — COMMENTED** [review]
No major comments. Feel free to update them in other commits.

File: oap-ape/ape-java/ape-hcfs/src/main/java/com/intel/oap/fs/hadoop/ape/hcfs/redis/RedisClient.java [hunk 4]
```diff
@@ -0,0 +1,371 @@
+package com.intel.oap.fs.hadoop.ape.hcfs.redis;
```
Comment (jikunshang) [inline oap-ape/ape-java/ape-hcfs/src/main/java/com/intel/oap/fs/hadoop/ape/hcfs/redis/RedisClient.java:1]: Please add license.

Comment (iyupeng) [reply oap-ape/ape-java/ape-hcfs/src/main/java/com/intel/oap/fs/hadoop/ape/hcfs/redis/RedisClient.java:1]: added in pull request #26

**Pull request merged**